ros/src/twist_controller/twist_controller.py

```diff
@@ -6,65 +6,111 @@
 GAS_DENSITY = 2.858
 ONE_MPH = 0.44704
 
+PID_SPDCTL_P = 0.9
+PID_SPDCTL_I = 0.0009
+PID_SPDCTL_D = 1.75
+
+
+PID_ACC_P = 0.4
+PID_ACC_I = 0.05
+PID_ACC_D = 0.4
+
+LPF_ACCEL_TAU = 0.2
 
 class Controller(object):
-    def __init__(self,
+    def __init__(self, sampling_rate,
                  vehicle_mass, fuel_capacity,
                  brake_deadband, decel_limit, accel_limit, wheel_radius,
                  wheel_base, steer_ratio, max_lat_accel, max_steer_angle):
-        
-        self.yaw_controller = YawController(wheel_base, steer_ratio, 0.1, max_lat_accel, max_steer_angle)
-
-        kp = 0.3
-        ki = 0.1
-        kd = 0.
-        mn = 0.   # max throttle
-        mx = 0.2  # min throttle
-        self.throttle_controller = PID(kp, ki, kd, mn, mx)
-
-        tau = 0.5 # 1/(2pi*tau) = cutoff frequency
-        ts = 0.02 # Sample time
-        self.vel_lpf = LowPassFilter(tau, ts)
 
         self.vehicle_mass = vehicle_mass
         self.fuel_capacity = fuel_capacity
         self.brake_deadband = brake_deadband
         self.decel_limit = decel_limit
         self.accel_limit = accel_limit
         self.wheel_radius = wheel_radius
+        self.sampling_rate = sampling_rate
+
+        tau = LPF_ACCEL_TAU # 1/(2pi*tau) = cutoff frequency
+        ts = 1./self.sampling_rate # Sample time
+
+        # rospy.logwarn('TwistController: Sampling rate = ' + str(self.sampling_rate))
+        # rospy.logwarn('TwistController: ts  = ' + str(ts))
+
+        self.prev_vel = 0.0
+        self.current_accel = 0.0
+        self.acc_lpf = LowPassFilter(tau, ts)
+
+        self.brake_torque_const = (self.vehicle_mass + self.fuel_capacity * GAS_DENSITY) * self.wheel_radius
 
-        self.last_time = rospy.get_time()
+        # Initialise PID controller for speed control
+        self.pid_spd_ctl = PID(PID_SPDCTL_P, PID_SPDCTL_I, PID_SPDCTL_D,
+                                  self.decel_limit, self.accel_limit)
+
+        # second controller to get throttle signal between 0 and 1
+        self.accel_pid = PID(PID_ACC_P, PID_ACC_I, PID_ACC_D, 0.0, 0.8)
+
+        # Initialise Yaw controller for steering control
+        self.yaw_controller = YawController(wheel_base, steer_ratio, 0.1, max_lat_accel, max_steer_angle)
+
+        # self.last_time = rospy.get_time()
 
     def reset(self):
-        self.throttle_controller.reset()
+        #self.accel_pid.reset()
+        self.pid_spd_ctl.reset()
 
     def control(self, current_vel, dbw_enabled, linear_vel, angular_vel):
         """Returns (throttle, brake, steer), or None"""
 
+        throttle, brake, steering = 0.0, 0.0, 0.0
+
         if not dbw_enabled:
             self.reset()
             return None
 
-        current_vel = self.vel_lpf.filt(current_vel)
-
-        steering = self.yaw_controller.get_steering(linear_vel, angular_vel, current_vel)
-
+        sampling_time = 1./self.sampling_rate
```

Comment:
this and `ts = 1./self.sampling_rate` above can be saved as an instance variable